[docs] Add documents for KoP's Schema Registry (#1684)
streamnative/kop: docs/schema.md

# Schema Registry

KoP has implemented the Schema Registry to support Kafka clients with [Confluent's Schema serializers and deserializers](https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#supported-formats).

> To enable the schema registry, you should add the following configuration:

Author: Just like other KoP configurations. If you start a broker with KoP, add it to `broker.conf`. If you start a standalone with KoP, add it to `standalone.conf`. So I think it would be redundant to mention which `.conf` file should it be added to.

```properties
kopSchemaRegistryEnable=true
```

The Schema Registry listens on the 8001 port by default, which means you should add the following property to create your Kafka producer or consumer.

```properties
schema.registry.url=http://<broker-ip>:8001
```

> **Note**
>

> > See [here](./configuration.md) for other configurations, like how to change the port it listens on.

Reviewer: ```suggestion
> For other configurations, such as how to change the port where the Schema Registry service listens on, see the [configuration guide](./configuration.md).
```

To manage the schemas, see the [Confluent's Schema REST API](https://docs.confluent.io/platform/current/schema-registry/develop/api.html#compatibility).

> ## Example: Use KoP Schema Registry with Confluent's Avro serializer

Reviewer: It seems that your example contains the following steps:

1. Enable the Schema Registry service on KoP
2. Define the Avro schema
3. Basic end-to-end example with Confluent's Avro serializer
4. Query the created schema

Among these 4 steps, the 3rd one is not a task title (I mean using a verb + noun format). It would be better to update the title for step 3. Therefore, these 4 sections keep the same format.

> This section adds an example to show how to use Confluent's Avro serializer on KoP.

Reviewer: ```suggestion
This section provides an example about how to use Confluent's Avro serializer on KoP.
```

> ### Enable Schema Registry in KoP standalone

Reviewer: ```suggestion
### Enable the Schema Registry service on KoP
```

> Follow the **Set configuration for KoP** section in [here](./kop.md), start the KoP standalone with the following configurations:

Reviewer: ```suggestion
Start KoP with the following properties in standalone mode. For details, see [Set configuration for KoP](./kop.md#set-configuration-for-kop).
```

```properties
messagingProtocols=kafka
kafkaListeners=PLAINTEXT://0.0.0.0:9092
brokerEntryMetadataInterceptors=org.apache.pulsar.common.intercept.AppendIndexMetadataInterceptor
brokerDeleteInactiveTopicsEnabled=false
allowAutoTopicCreationType=partitioned

# Enable the transaction to be compatible with Kafka clients 3.2.0 or later
# See https://kafka.apache.org/documentation/#upgrade_320_notable
kafkaTransactionCoordinatorEnabled=true
brokerDeduplicationEnabled=true

kopSchemaRegistryEnable=true
```

### Define the Avro schema

> Create an Avro schema file named `User.asvc` under the `src/main/avro` directory:

Reviewer: Does this sentence and the sentence below has a sequential order? If yes, i would like to suggesting using an steplist format. Otherwise, ignore my comment.

Author: No.

```json
{
  "namespace": "example.avro",
  "type": "record",
  "name": "User1",
  "fields": [
    {"name": "name", "type": "string"},
    {"name": "age", "type": "int"}
  ]
}
```

> Add the `avro-maven-plugin` in order to generate the Java class from the Avro Schema file above.

Reviewer: ```suggestion
Add the `avro-maven-plugin` plugin to generate the Java class from the Avro schema file above.
```

```xml
      <plugin>
        <groupId>org.apache.avro</groupId>
        <artifactId>avro-maven-plugin</artifactId>
        <version>1.10.2</version>
        <executions>
          <execution>
            <phase>generate-sources</phase>
            <goals>
              <goal>schema</goal>
            </goals>
            <configuration>
              <sourceDirectory>${project.basedir}/src/main/avro</sourceDirectory>
              <includes>
                <include>User.avsc</include>
              </includes>
              <outputDirectory>${project.build.directory}/generated-sources</outputDirectory>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

With the Maven plugin above, a `User` class that has a `String` field named `name` and an `int` field named `age` will be generated under the `example.avro` package. 

### Basic end-to-end example with Confluent's Avro serializer

> Add the following dependencies to use Kafka client 3.3.1 with Confluent's Avro serializer 7.3.1:

Reviewer: same comment as in the above section.

```
Sent to my-avro-kafka-topic-0@0
```

Run the following consumer application:

```java
final Properties props = new Properties();
props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, KafkaAvroDeserializer.class);
props.put(ConsumerConfig.GROUP_ID_CONFIG, "my-group");
props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
props.put(KafkaAvroSerializerConfig.SCHEMA_REGISTRY_URL_CONFIG, "http://localhost:8001");

final KafkaConsumer<String, User> consumer = new KafkaConsumer<>(props);
consumer.subscribe(Collections.singleton("my-avro-kafka-topic"));
while (true) {
    ConsumerRecords<String, User> records = consumer.poll(Duration.ofSeconds(1));
    records.forEach(record -> System.out.println("Received " + record.value() + " from "
            + record.topic() + "-" + record.partition() + "@" + record.offset()));
    if (!records.isEmpty()) break;
}
consumer.close();
```

You will see the following output:

```
Received {"name": "alice", "age": 10} from my-avro-kafka-topic-0@0
```

### Query the created schema

> Query the schema whose id is 1 (i.e. the 1st schema):

Reviewer: ```suggestion
Query the schema whose ID is 1 (i.e. the 1st schema):
```

```bash
$ curl -L http://localhost:8001/schemas/ids/1; echo
{
  "schema" : "{\"type\":\"record\",\"name\":\"User\",\"namespace\":\"example.avro\",\"fields\":[{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"age\",\"type\":\"int\"}]}"
}
```

Query the subjects:

```bash
$ curl -L http://localhost:8001/subjects; echo
[ "my-avro-kafka-topic-value" ]
$ curl -L http://localhost:8001/subjects/my-avro-kafka-topic-value/versions; echo
[ 1 ]
$ curl -L http://localhost:8001/subjects/my-avro-kafka-topic-value/versions/1; echo
{
  "id" : 1,
  "schema" : "{\"type\":\"record\",\"name\":\"User\",\"namespace\":\"example.avro\",\"fields\":[{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"age\",\"type\":\"int\"}]}",
  "subject" : "my-avro-kafka-topic-value",
  "version" : 1,
  "type" : "AVRO"
}
```

> See [here](https://docs.confluent.io/platform/current/schema-registry/schema_registry_tutorial.html#terminology-review) for a quick start of Confluent's Schema concepts.

Reviewer: ```suggestion
For a quick start of Confluent's Schema concepts, See [Confluent documentation](https://docs.confluent.io/platform/current/schema-registry/schema_registry_tutorial.html#terminology-review) .
```
If this sentence belongs to the query created schema section, it is OK to put the sentence here. If it is just a reference that you want users to get more info, it would be great to use a format, such as:
## References
For a quick start of Confluent's Schema concepts, See [Confluent documentation](https://docs.confluent.io/platform/current/schema-registry/schema_registry_tutorial.html#terminology-review) .
